archive/imo/imo1994_q1.lean

```diff
@@ -0,0 +1,105 @@
+
+# Sketch of solution
+
+We can order the numbers so that `a₁ ≤ a₂ ≤ ... ≤ aₘ`.
+The key idea is to pair the numbers in the sum and show that `aᵢ + aₘ₊₁₋ᵢ ≥ n+1`.
+Indeed, if we had `aᵢ + aₘ₊₁₋ᵢ ≤ n`, then `a₁ + aₘ₊₁₋ᵢ, a₂ + aₘ₊₁₋ᵢ, ..., aᵢ + aₘ₊₁₋ᵢ`
+would be `m` elements of the set of `aᵢ`'s all larger than `aₘ₊₁₋ᵢ`, which is impossible.
+-/
+
+open_locale big_operators
+
+open finset
+
+lemma tedious (m : ℕ) (k : fin (m+1)) : m - (m + (m + 1 - ↑k)) % (m + 1) = ↑k  :=
+begin
+  cases k with k hk,
+  simp only [fin.coe_mk],
+  rw [nat.lt_succ_iff,le_iff_exists_add] at hk,
+  rcases hk with ⟨c, rfl⟩,
+  have foo : k + c + 1 - k = c + 1,
+    rw [add_assoc, add_tsub_cancel_left],
+  rw foo,
+  rw (show k + c + (c + 1) = c + (k + c + 1), by ring),
+  simp only [nat.add_mod_right],
+  rw nat.mod_eq_of_lt (show c < k + c + 1, by linarith),
+  simp,
+end
+
+theorem imo1994_q1 (n : ℕ) (m : ℕ) (A : finset ℕ) (hm : A.card = m + 1)
+  (hrange : ∀ a ∈ A, 0 < a ∧ a ≤ n) (hadd : ∀ (a b ∈ A), a + b ≤ n → a + b ∈ A) :
+  (m+1)*(n+1) ≤ 2*(∑ x in A, x) :=
+begin
+  set a := order_emb_of_fin A hm,  -- We sort the elements of A
+  set rev := equiv.sub_left (fin.last m), -- i ↦ m-i
+
+  -- The key inequality
+  have hpair : ∀ k ∈ univ, (a k) + a (rev k) ≥ n+1,
+  { rintros k -,
+    by_contradiction,
+    --simp [nat.lt_succ_iff] at h,
+
+    -- We exhibit k+1 elements of A greater than a (rev k)
+    set f : fin (m+1) ↪ ℕ := ⟨λ i, a i + a (rev k),
+    begin
+      apply injective_of_le_imp_le,
+      intros i j hij,
+      rwa [add_le_add_iff_right, a.map_rel_iff] at hij,
+    end⟩,
+
+    -- Proof that the f i are greater than a (rev k) for i≤k
+    have hf : map f (Icc 0 k) ⊆ map a.to_embedding (Ioc (rev k) (fin.last m)),
+    { intros x hx,
+      simp at *,
+      rcases hx with ⟨i, ⟨hi, rfl⟩⟩,
+      have h1 : a i + a (fin.last m - k) ≤ n,
+      { suffices hik : a i ≤ a k, linarith,
+        exact a.monotone hi.2 },
+      have h2 : a i + a (fin.last m - k) ∈ A :=
+        hadd _ _ (order_emb_of_fin_mem A hm _) (order_emb_of_fin_mem A hm _) h1,
+      rw [←mem_coe, ←range_order_emb_of_fin A hm, set.mem_range] at h2,
+      cases h2 with j hj,
+      use j,
+      split, split, rotate 1,
+      { exact fin.le_last j },
+      { exact hj },
+      rw [←strict_mono.lt_iff_lt a.strict_mono, hj],
+      simp,
+      exact (hrange (a i) (order_emb_of_fin_mem A hm i)).1 },
+
+    -- A set of size k+1 embed in one of size k, which yeilds a contradiction
+    have ineq := card_le_of_subset hf,
+    simp [fin.coe_sub, tedious] at ineq,
+    contradiction },
+
+  -- We reindex the sum by fin (m+1)
+  have : ∑ x in A, x = ∑ i : fin (m+1), a i,
+  { convert sum_image (λ x hx y hy, (order_embedding.eq_iff_eq a).1),
+    rw ←coe_inj, simp },
+  rw this, clear this,
+
+  -- The rest is a simple calculation by rearranging one of the two sums
+  calc 2 * ∑ i : fin (m+1), a i = ∑ i : fin (m+1), a i + ∑ i : fin (m+1), a i       : two_mul _
```

Comment:
```suggestion
  calc 2 * ∑ i : fin (m+1), a i
      = ∑ i : fin (m+1), a i + ∑ i : fin (m+1), a i       : two_mul _
```
Please reformat the other lines as well. Hopefully they'll fit in ≤ 100 chars that way.

Reply:
Thanks, done!